napari/layers/shapes/shapes.py

```diff
@@ -2006,7 +2006,7 @@ def add(
                 face_color=face_color,
                 z_index=z_index,
             )
-            self.events.data(value=self.data)
+            self.events.data(value=self.data, action="add", data_indices=[-1])
```

Comment:
why `[-1]` instead of `len(self.data)-1`?

Reply:
Simply because when adding the shape gets added at the last index. Is there a particular reason for going with `len(self.data)-1`?

Reply:
Because the assumption that all coordinates are positive simplifies many calculations. But I do not have a strong opinion here.

Reply:
If a `frozenset` is used, doesn't that ruin the ability to enumerate over `data_indices` and `vertex_indices`? When it was a list we could rely on `vertex_indices[i]` corresponding to the shape at `data_indices[i]` - since `frozenset` is unordered we can't do that anymore right?

Reply:
That is true when trying to do this directly. one can cast it to a list though. It is a bit inconvenient, but doable. Question is what value other than semantic correctness it provides then I think and whether we choose for that or ease of use.

Reply:
Even if I choose to cast it as a list later, am I not still losing the correspondence between indices? I don't see how you could get the mapping back once the data has been placed into two separate sets

Reply:
Even if I choose to cast it as a list later, am I not still losing the correspondence between indices? I don't see how you could get the mapping back once the data has been placed into two separate sets

Reply:
ah yes we lose order. I will revert back

Reply:
@Czaki We thought of using a frozenlist, but these can't be nested. Another would be to use tuples, but before going that direction, is there any particular concern that you have with lists?

Reply:
@melonora This value is shared between all functions that receive events, so every prevention that one handler pollutes data for another handler. (frozenlist is tuple)

Reply:
@Czaki I opted for tuples as they can be created in a nested manner. Think that is the best we can do. Are you ok with this?